File: python/longestSubstringWithoutRepeatingCharacters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if len(s) < 2:
            return len(s)
        left, right, longest = 0, 1, 0
        charset = {s[left]}
        while right < len(s):
            if s[right] not in charset:
                charset.add(s[right])
                longest = max(longest, right - left)
                right += 1
            else:
                charset.remove(s[left])
                left += 1
        return longest + 1
        
        

File: javascript/longestSubstringWithoutRepeatingCharacters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    if (s.length < 2) {
        return s.length;
    }
    let left = 0, right = 1, max = 0;
    let set = new Set();
    set.add(s[left]);
    while (right < s.length) {
        if (!set.has(s[right])) {
            set.add(s[right]);
            max = Math.max(right - left, max);
            right++;
        } else {
            set.delete(s[left]);
            left++;
        }
    }
    return max + 1;
};
